### PR TITLE
Implement #507

### LINKED ISSUE #507: used_underscore can't handle macros and derive properly at the same time
I removed the test in https://github.com/Manishearth/rust-clippy/commit/4a32445aa7fdc7d38fdfeb23e07536ff11c860bc because we need a stronger macro check to handle stuff like https://github.com/Manishearth/rust-clippy/issues/488#issuecomment-166032639

Might need to write a better macro check. Looks like hygiene works differently for `MacroAttribute`s. Perhaps we should just check for those?

cc @devonhollowood

### PR DESCRIPTION
**Pull request by devonhollowood**
Make `used_underscore_binding` lint compatible with MacroAttributes
expansions. TODO: Add a good test for this.

Fixes #507

### CODE REVIEW

**Comment by Manishearth** [comment]
Did you check that this works with rustc-serialize?

(You can add a regular test similar to mut_mut_macro that does this)

File: tests/compile-fail/used_underscore_binding.rs [hunk 1]
```diff
@@ -9,7 +9,15 @@ fn prefix_underscore(_foo: u32) -> u32 {
 
 /// Test that we lint even if the use is within a macro expansion
 fn in_macro(_foo: u32) {
-    println!("{}", _foo); // doesn't warn, nut should #507
+    println!("{}", _foo); //~ ERROR used binding which is prefixed with an underscore
+}
+
+// TODO: This doesn't actually correctly test this. Need to find a #[derive(...)] which sets off
```
Comment (Manishearth) [inline tests/compile-fail/used_underscore_binding.rs]: `#[derive(RustcEncodable)]` should work. However, you need to include `rustc_serialize`, which is a Crates.io crate, so it will not work directly in a compile-fail test. Instead, add a small test, similar to mut_mut_macro, directly in `tests/` (which uses RustcEncodable on a non-empty struct and has `#[deny(this lint)]`), and a dev-dependency on rustc-serialize in Cargo.toml. The test should always pass since this lint will not trigger there.

(Ensure the test fails when you use it for an earlier clippy version)

Comment (Manishearth) [inline tests/compile-fail/used_underscore_binding.rs]: (see also: https://travis-ci.org/untitaker/mysteryshack/jobs/97893563#L305; https://github.com/untitaker/mysteryshack/blob/master/src/mysteryshack/web/oauth.rs#L21)

**Comment by devonhollowood** [comment]
Done. I checked, and it does not compile (because of the `deny`) for the current HEAD on master (4a32445aa7fdc7d38fdfeb23e07536ff11c860bc) if I delete the early return that you added for the lint.

**Comment by Manishearth** [comment]
Awesome, thanks!
